### PR TITLE
Add a skip field if a trailing delimiter is detected.

### LINKED ISSUE #22: Trailing Delimiters (or a mapping must have a field and not a delimiter as the last element)
More of  question, but I when I was testing replacing my apache combined/common grok lines with dissect I had to add an empty match %{} onto the end of the mapping otherwise the last double quote (") would appear in the useragent. Is this the desired behaviour for dissect?

Mapping
%{clientIP} %{rfcIdent} %{userName} [%{timestamp} %{+timestamp}] "%{method} %{uriStem} HTTP/%{httpVersion}" %{responseCode} %{bytesRecv} "%{referrer}" "%{userAgent}"%{}

### PR DESCRIPTION
**Pull request by guyboertje**
Increased stricter inforcing of delimiter find sequence and bail early.
More strict on handling leading delimiters, they must be found at
the beginning of the value.
Add the compatibility tests from Beats.
Some (OK quite a few) IntelliJ inspection changes.
non-inspection changes are restricted to:
Dissector.java
DissectorTest.java
DissectResult.java
Field.java
FieldFactory.java
JavaDissectorLibrary.java
dissect_spec.rb
plus the VERSION and CHANGELOG.md

Fixes #22

### CODE REVIEW

**Comment by guyboertje** [comment]
No significant change in JMH benchmarks.

File: src/main/java/org/logstash/dissect/fields/Field.java [hunk 1]
```diff
@@ -13,17 +13,24 @@
     int NORMAL_ORDINAL_LOWER = 1;
     int APPEND_ORDINAL_BASE = 100;
     int INDIRECT_ORDINAL_HIGHER = 1000;
-    int MISSING_ORDINAL_HIGHEST = 100000;
+    // int MISSING_ORDINAL_HIGHEST = 100000;
     Pattern SUFFIX_REGEX = Pattern.compile("(.+?)(/\\d{1,2}|->|/\\d{1,2}->|->/\\d{1,2})?$");
+    Pattern SUFFIX_ONLY_REGEX = Pattern.compile("^(/\\d{1,2}|->|/\\d{1,2}->|->/\\d{1,2})?$");
```
Comment (robbavey) [inline src/main/java/org/logstash/dissect/fields/Field.java]: Consider making these Patterns `static`

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+## 1.2.0
+  - Fix Trailing Delimiters requires a false field. A skip field is
+  automatically added when a final delimiter is detected in the dissect pattern.
+  This requires that strict delimiter finding is enforced  - meaning a "no match"
+  results if every delimiter is not found in exactly the declared order
```
[the remark this answers is not in the document]
Comment (guyboertje) [reply CHANGELOG.md]: One can argue that it is a bug fix to an expectation mismatch.
We say in the docs that Dissect scans for a set of delimiters from left to right - the user is made to understand that this is strict but we weren't.

We should discuss this.

If it is considered a breaking change then we can't release it until 7.0.0 because the plugin will have be at 2.0.0. This seems quite artificial to me because after 7.0.0 is released a 6.X user can update and not have the opportunity to see a breaking changes warning.

File: src/main/java/org/logstash/dissect/fields/Field.java (hunk 1 above)
Comment (guyboertje) [reply src/main/java/org/logstash/dissect/fields/Field.java]: TIL "All the fields declared in an interface are implicitly public, static and final constants"

File: CHANGELOG.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsvd) [reply CHANGELOG.md]: > If it is considered a breaking change then we can't release it until 7.0.0 because the plugin will have be at 2.0.0

**If** the breaking change is necessary we can release it now but we should fork the current major into a branch so we can continue to release bugfixes/features in 1.x

> This seems quite artificial to me because after 7.0.0 is released a 6.X user can update and not have the opportunity to see a breaking changes warning.

The update process will show that a plugin was upgraded to a new major, at that point something will likely break. Ideally the plugin manager would display any breaking change notice, but that's not possible right now.

[the remark this answers is not in the document]
Comment (guyboertje) [reply CHANGELOG.md]: One question, what is our definition of a breaking change?
1. This change will break every installation.
2. This change might break a some installations (identifiable).
3. This change might break a very small number of installations (unidentifiable).

[the remark this answers is not in the document]
Comment (guyboertje) [reply CHANGELOG.md]: @jsvd and I decided that this is not a breaking change because the docs are explicit about having delimiters in between fields and it is not specific about the behaviour when a delimiter is not found in a particular place.

[the remark this answers is not in the document]
Comment (robbavey) [reply CHANGELOG.md]: 👍
